### DISCUSSION TITLE
Including data-gjs-type in exported html

### DISCUSSION THREAD

**Discussion by benjgrad** [GrapesJS/grapesjs#4518] [Ideas]
## Problem
As mentioned in[ issue 324](https://github.com/artf/grapesjs/issues/324), users that want to edit source HTML will see that the props are removed from the HTML, meaning that upon save, the new document will not recognize components.
 ## Solution: withProps option
In components.md, a flag is documented to including component properties as data-gjs-* attributes. We can use this same code by providing a global config to allow the necessary data to be preserved. The flag should default as false to preserve existing behaviour.

https://github.com/benjgrad/grapesjs/commit/9e00994fd278120a366e528cf35aa043fdf2ceca

**Comment by benjgrad** [comment]
(I tried pushing this to a branch in the grapesjs repo, but I got permission denied. Also, I don't see any way of adding Labels to this... not sure if it's related to permissions)

**Comment by artf** [comment]
You should be able to pass that option as default already by using `optsHtml` on editor init:
https://github.com/artf/grapesjs/blob/054e352172e6580efd7859c2121a858187597d1b/src/editor/config/config.ts#L103-L104

**Reply by benjgrad** [reply]
I think the issues is it’s not being passed down to the HTMLGenerator then… is this a bug?

**Reply by benjgrad** [reply]
I also added `withProps` to the docs, since the only 2 options are `opts.component` and `opts.cleanId`.

**Reply by denny99** [reply]
Kinda funny. I am struggling with the same issue here.

<img width="432" alt="image" src="https://user-images.githubusercontent.com/1826849/185906087-f375963c-c3fc-4a10-8d61-d0fd4c54e4b2.png">

I am creating my own preview dialog with monaco so the user can edit the html directly. To reimport the html properly i wanted to use this `withProps` flag but as you can see in the code above, i have no chance of passing this to the htmlOpts on demand. The function creates the htmlOpts from scratch and just uses the two properties of opts that are described in the docs

**Reply by denny99** [reply]
My somewhat hacky solution is just to bypass the original getHtml function and use the internal methods

`editor.CodeManager.getGenerator('html').buildModels(editor.getComponents(), {withProps: true})`

**Reply by benjgrad** [reply]
@artf any thoughts? It's looking like this may be a bug.

I have a demo of the issue with your suggested changes here:

https://www.gradyserver.com/issue

A build of the changes I linked above can be found here:

https://www.gradyserver.com/edit

**Reply by artf** [reply]
Yeah @denny99 is right, thanks for checking this out, I'll fix it in the next release.
